src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -110,44 +121,126 @@ protected override void Dispose(bool disposing)
+            IntPtr headerPtr = IntPtr.Zero;
+            bool referenceTaken = false;
+
+#if DEBUG
+            if (Logging.On) Logging.Enter(Logging.Http, this, "StartRequest", state.RequestMessage);
+#endif
 
             if (state.CancellationToken.IsCancellationRequested)
             {
-                state.TrySetCanceled();
+                state.TrySetCanceled(state.CancellationToken);
                 return;
             }
 
-            var cancellationTokenRegistration = state.CancellationToken.Register(x => ((RequestCompletionSource) x).TrySetCanceled(), state);
+            var cancellationTokenRegistration = state.CancellationToken.Register(
+                s =>
+                {
+                    var rcs = (RequestCompletionSource)s;
+                    rcs.TrySetCanceled(rcs.CancellationToken);
+                },
+                state);
 
             try
             {
-                requestHandle = CreateRequestHandle(state);
+                // Prepare context objects
+                requestStateHandle = GCHandle.Alloc(state);
+                responseMessage = new CurlResponseMessage(state.RequestMessage);
+                state.ResponseMessage = responseMessage;
+                requestHandle = CreateRequestHandle(state, requestStateHandle, out headerPtr);
 
-                state.CancellationToken.ThrowIfCancellationRequested();
-                int result = Interop.libcurl.curl_easy_perform(requestHandle);
-                if (CURLcode.CURLE_OK != result)
+                if (state.CancellationToken.IsCancellationRequested)
                 {
-                    throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+                    state.TrySetCanceled(state.CancellationToken);
+                    return;
                 }
 
-                // TODO: Handle requests with body
+                if (state.RequestMessage.Content != null)
+                {
+#if DEBUG
+                    Logging.PrintInfo(Logging.Http, this, "Request content for " + Logging.GetObjectLogHash(responseMessage.ContentStream) + " : " + state.RequestMessage.Content.ReadAsStringAsync().Result);
+#endif
+
+                    Stream requestContentStream =
+                        await state.RequestMessage.Content.ReadAsStreamAsync().ConfigureAwait(false);
+                    if (state.CancellationToken.IsCancellationRequested)
+                    {
+                        state.TrySetCanceled(state.CancellationToken);
+                        return;
+                    }
+                    state.RequestContentStream = requestContentStream;
+                }
 
-                state.CancellationToken.ThrowIfCancellationRequested();
-                responseMessage = CreateResponseMessage(requestHandle, state.RequestMessage);
+                requestHandle.DangerousAddRef(ref referenceTaken);
+                var performTask = Task.Run(() =>
```

Comment:
I'm not clear on the interactions with libcurl.  From seeing some of the callbacks defined earlier, I was thinking that libcurl provided an asynchronous programming model, but curl_easy_perform as used below and the associated comments imply that curl_easy_perform is actually a synchronous method, such that it doesn't return until all work has completed?  Is there an async version of it?  We should get very crisp on what the async nature of what's being used is:
- If the underlying library only exposes synchronous entry points, such that we call a method like curl_easy_perform and it doesn't return until all work has completed, all data has been transferred, etc., then I'm not sure why we're using async/await and TaskCompletionSource at all: SendAsync would just use StartNew to schedule a task that performs all of the work synchronously.
- If the underlying library does expose asynchronous entry points, such that there's some function we can call which kicks off the async I/O and then we get callbacks when the work has completed, then we shouldn't need to be using either StartNew or Run: we call the async operation, and we give it a callback that completes the TaskCompletionSource.

As currently written, this seems to be a mishmash of both, incurring the downsides of both.

If nothing else, I don't understand why we have both StartNew and Run being used as part of a single operation.  And the use of Run here is resulting in a delegate/closure.